src/server/courseInstance/courseInstance.controller.ts

```diff
@@ -97,9 +104,58 @@ export class CourseInstanceController {
   @Get('/multi-year-plan')
   public async getMultiYearPlan(): Promise<MultiYearPlanResponseDTO[]> {
     // The number of years specified for the multi year plan
-    const numYears = 4;
+    const numYears = CourseInstanceController.NUM_YEARS;
+    const expectedNumSemesters = CourseInstanceController.NUM_SEMESTERS;
     const academicYears = this.computeAcademicYears(numYears);
-    return this.ciService.getMultiYearPlan(academicYears);
+    let multiYearPlans
+    : MultiYearPlanResponseDTO[] = await this.ciService
+      .getMultiYearPlan(academicYears);
+    const numSemestersReturned = multiYearPlans[0]
+      ? multiYearPlans[0].semesters.length
+      : 0;
+    const missingNumSemesters = expectedNumSemesters - numSemestersReturned;
+    // In cases in which the database does not have sufficient semester rows, we
+    // would like to populate the missing semesters so that we always get the
+    // number of years worth of multi-year plans that is requested.
+    const createMissingSemester = (offset: number) => {
+      const semesterIndex = numSemestersReturned + offset;
+      // E.g. If the first academic year is 2021, numSemestersReturned is 4
+      // and offset is 0, then
+      // firstMissingAcademicYear = 2021 + Math.floor((4 + 0) / 2) = 2023,
+      // and term will be FALL.
+      // If offset is 1, then firstMissingAcademicYear is still 2023,
+      // while term will be SPRING.
+      const firstMissingAcademicYear = academicYears[0]
+        + Math.floor(semesterIndex / 2);
+      const term = semesterIndex % 2 === 0
+        ? TERM.FALL
+        : TERM.SPRING;
+      const calendarYear = term === TERM.FALL
+        ? firstMissingAcademicYear - 1
+        : firstMissingAcademicYear;
+      return {
+        id: `missingSemester${offset}`,
+        academicYear: String(firstMissingAcademicYear),
+        calendarYear: String(calendarYear),
+        term,
+        instance: {
+          id: `missingInstance${offset}`,
+          faculty: [],
+        },
+      };
+    };
+    const missingSemesters: MultiYearPlanSemester[] = Array.from(
+      new Array(missingNumSemesters)
+    ).map((_, index) => (
+      createMissingSemester(index)
+    ));
+    if (missingNumSemesters > 0) {
+      multiYearPlans = multiYearPlans.map((plan) => ({
+        ...plan,
+        semesters: [...plan.semesters, ...missingSemesters],
+      }));
```

Comment:
@jonseitz Thanks for this suggestion! I think this makes a lot of sense. I reverted the controller logic back to return what is available in the db in commit 830ff75 along with updating the client to use the constants in the `common/constants` directory to calculate the right number of columns/filters.

I noticed the build is not running in this PR - does having a merge conflict prevent the build from running? Should I try to resolve the merge conflict now before re-review?

Reply:
Ah, so the builds aren't running because the workflow files don't exist in this branch yet. You'd need to merge develop back into here before those will run.

Reply:
D'oh! Thanks Jonathan.